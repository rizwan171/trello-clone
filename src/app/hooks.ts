import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";
import type { RootState, AppDispatch } from "./store";

// The below useAppDispatch does not require a return type, hence the eslint disable comment
// See https://stackoverflow.com/questions/70711795/missing-return-type-on-function-useappdispatch for more info
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
