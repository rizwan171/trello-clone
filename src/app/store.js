import { configureStore } from "@reduxjs/toolkit";
import boardReducer from "../features/boardSlice.js";
import listsReducer from "../features/listsSlice.js";
import cardsReducer from "../features/cardsSlice.js";
import boardOptionsReducer from "../features/boardOptionSlice.js";
import currentSelectedCardReducer from "../features/currentSelectedCardSlice.js";
import currentSelectedListReducer from "../features/currentSelectedListSlice.js";
import listOptionsMenuPositionReducer from "../features/listOptionsMenuPositionSlice.js";
import tagsReducer from "../features/tagsSlice.js";
import selectedTagColourReducer from "../features/selectedTagColourSlice.js";
import createTagMenuDataReducer from "../features/createTagMenuDataSlice.js";
import modalActionMenusVisibilityReducer from "../features/modalActionMenusVisibilitySlice.js";

export const store = configureStore({
  reducer: {
    board: boardReducer,
    lists: listsReducer,
    cards: cardsReducer,
    boardOptions: boardOptionsReducer,
    currentSelectedCard: currentSelectedCardReducer,
    currentSelectedList: currentSelectedListReducer,
    listOptionsMenuPosition: listOptionsMenuPositionReducer,
    tags: tagsReducer,
    selectedTagColour: selectedTagColourReducer,
    createTagMenuData: createTagMenuDataReducer,
    modalActionMenusVisibility: modalActionMenusVisibilityReducer,
  },
});
