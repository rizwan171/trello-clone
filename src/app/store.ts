import { combineReducers, configureStore, PreloadedState } from "@reduxjs/toolkit";
import boardReducer from "../features/boardSlice";
import listsReducer from "../features/listsSlice";
import cardsReducer from "../features/cardsSlice";
import boardOptionsReducer from "../features/boardOptionSlice";
import currentSelectedCardReducer from "../features/currentSelectedCardSlice";
import currentSelectedListReducer from "../features/currentSelectedListSlice";
import listOptionsMenuPositionReducer from "../features/listOptionsMenuPositionSlice";
import tagsReducer from "../features/tagsSlice";
import selectedTagColourReducer from "../features/selectedTagColourSlice";
import createTagMenuDataReducer from "../features/createTagMenuDataSlice";
import modalActionMenusVisibilityReducer from "../features/modalActionMenusVisibilitySlice";

const rootReducer = combineReducers({
  board: boardReducer,
  lists: listsReducer,
  cards: cardsReducer,
  boardOptions: boardOptionsReducer,
  currentSelectedCard: currentSelectedCardReducer,
  currentSelectedList: currentSelectedListReducer,
  listOptionsMenuPosition: listOptionsMenuPositionReducer,
  tags: tagsReducer,
  selectedTagColour: selectedTagColourReducer,
  createTagMenuData: createTagMenuDataReducer,
  modalActionMenusVisibility: modalActionMenusVisibilityReducer,
});

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};

export type RootState = ReturnType<typeof rootReducer>
export type AppDispatch = AppStore['dispatch']
export type AppStore = ReturnType<typeof setupStore>