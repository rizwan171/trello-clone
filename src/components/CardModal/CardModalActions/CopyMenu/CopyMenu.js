import React from "react";
import { MdClose } from "react-icons/md";
import { useSelector } from "react-redux";

const CopyMenu = () => {
  const lists = useSelector((state) => state.lists.value);

  const handleClose = () => {};

  const handleCopyCard = () => {};

  return (
    <div className="fixed mt-16 w-72 text-gray-700 bg-white rounded-ibsm shadow-2xl p-4">
      <div className="relative text-center mb-2">
        <span className="text-sm block relative z-10">Copy Card</span>
        <MdClose onClick={handleClose} size={20} className="absolute right-0 top-0 z-20 cursor-pointer" />
      </div>
      <hr />
      <div className="flex flex-col mt-2 text-sm">
        <span className="text-gray-600 my-2 font-semibold">Copy To...</span>
        <div>
          <label htmlFor="lists absolute top-0 left-3">List</label>
          <select name="lists" className="rounded-sm p-2">
            {lists.map((list) => (
              <option value={list.id} key={list.id}>
                {list.title}
              </option>
            ))}
          </select>
        </div>
        <button
          onClick={handleCopyCard}
          className="p-2 mt-3 bg-trello-blue-100 hover:bg-trello-blue-200 text-white font-semibold items-center text-sm shadow-sm rounded-sm"
        >
          Copy Card
        </button>
      </div>
    </div>
  );
};

export default CopyMenu;
