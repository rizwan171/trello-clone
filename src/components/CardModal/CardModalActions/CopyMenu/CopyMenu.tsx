import { useState } from "react";
import { MdClose } from "react-icons/md";
import { useDispatch, useSelector } from "react-redux";
import { copyCardToList } from "../../../../features/cardsSlice";
import { closeMenu } from "../../../../features/modalActionMenusVisibilitySlice";

const CopyMenu = () => {
  const dispatch = useDispatch();
  const card = useSelector((state) => state.currentSelectedCard.value);
  const lists = useSelector((state) => state.lists.value);
  const [selectedList, setSelectedList] = useState("");

  const handleClose = () => {
    dispatch(closeMenu());
  };

  const handleCopyCard = () => {
    dispatch(copyCardToList({ id: card.id, listId: selectedList }));
    handleClose();
  };

  return (
    <div className="fixed mt-10 w-72 text-gray-700 bg-white rounded-ibsm shadow-2xl p-4">
      <div className="relative text-center mb-2">
        <span className="text-sm block relative z-10">Copy Card</span>
        <MdClose onClick={handleClose} size={20} className="absolute right-0 top-0 z-20 cursor-pointer" />
      </div>
      <hr />
      <div className="flex flex-col mt-2 text-sm">
        <span className="text-gray-600 my-2 font-semibold">Copy To...</span>
        <select
          name="lists"
          className="rounded-sm p-2 w-full max-w-full overflow-ellipsis"
          value={selectedList}
          onChange={(e) => setSelectedList(e.target.value)}
        >
          <option value="" selected disabled hidden>
            Select list
          </option>
          {lists.map((list) => (
            <option value={list.id} key={list.id}>
              {list.title}
            </option>
          ))}
        </select>
        <button
          onClick={handleCopyCard}
          className="p-2 mt-3 bg-trello-blue-100 hover:bg-trello-blue-200 text-white font-semibold items-center text-sm shadow-sm rounded-sm"
        >
          Copy Card
        </button>
      </div>
    </div>
  );
};

export default CopyMenu;
