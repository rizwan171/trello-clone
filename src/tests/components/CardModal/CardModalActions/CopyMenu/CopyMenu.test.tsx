import { fireEvent, screen, waitFor } from "@testing-library/react";
import { RootState } from "../../../../../app/store";
import CopyMenu from "../../../../../components/CardModal/CardModalActions/CopyMenu/CopyMenu";
import { renderWithProviders } from "../../../../utils/renderUtils";
import { generateModalActionsVisibilityState } from "../../../../utils/stateUtils";

describe("CopyMenu", () => {
  let initialState: Partial<RootState>;

  beforeEach(() => {
    const modalActionsVisibilityState = generateModalActionsVisibilityState();
    initialState = {
      cards: {
        value: [
          {
            id: "1",
            title: "Test Card",
            description: "Test Description",
            listId: "1",
            tags: [],
            attachments: [],
          },
        ],
      },
      lists: {
        value: [
          {
            id: "1",
            title: "List 1",
          },
          {
            id: "2",
            title: "List 2",
          },
        ],
      },
      currentSelectedCard: {
        value: {
          id: "1",
          title: "Test Card",
          description: "Test Description",
          listId: "1",
          tags: [],
          attachments: [],
        },
      },
      modalActionMenusVisibility: {
        value: {
          ...modalActionsVisibilityState.value,
          copyMenuOpen: true,
        },
      },
    };
  });

  it("should render successfully", () => {
    const view = renderWithProviders(<CopyMenu />, { preloadedState: initialState });
    expect(view.asFragment()).toMatchSnapshot();
  });

  it("should close copy menu", () => {
    const { store } = renderWithProviders(<CopyMenu />, { preloadedState: initialState });
    const stateBeforeClose = store.getState();
    expect(stateBeforeClose.modalActionMenusVisibility.value.copyMenuOpen).toBeTruthy();

    const closeIcon = screen.getByTestId("close-icon");
    expect(closeIcon).toBeInTheDocument();

    fireEvent.click(closeIcon);

    const stateAfterClose = store.getState();
    expect(stateAfterClose.modalActionMenusVisibility.value.copyMenuOpen).toBeFalsy();
  });

  it("should copy card to selected list", async () => {
    const { store } = renderWithProviders(<CopyMenu />, { preloadedState: initialState });
    const stateBeforeCopy = store.getState();
    expect(stateBeforeCopy.modalActionMenusVisibility.value.copyMenuOpen).toBeTruthy();
    expect(stateBeforeCopy.cards.value).toHaveLength(1);
    const card = stateBeforeCopy.cards.value[0];

    const listSelector = screen.getByRole("combobox") as HTMLSelectElement;
    expect(listSelector).toBeInTheDocument();
    expect(listSelector.options).toHaveLength(3);

    const expectedOptionValues = ["", "1", "2"];
    const actualOptionValues = [...listSelector.options].map((option) => option.value);
    expect(actualOptionValues).toStrictEqual(expectedOptionValues);

    const expectedTextValues = ["Select list", "List 1", "List 2"];
    const actualTextValues = [...listSelector.options].map((option) => option.textContent);
    expect(actualTextValues).toStrictEqual(expectedTextValues);

    fireEvent.change(listSelector, { target: { value: "2" } });

    const copyButton = screen.getAllByText("Copy Card").find((element) => element.tagName === "BUTTON") as HTMLButtonElement;
    expect(copyButton).toBeInTheDocument();

    fireEvent.click(copyButton);

    const stateAfterCopy = store.getState();
    expect(stateAfterCopy.cards.value).toHaveLength(2);
    const copiedCard = stateAfterCopy.cards.value[1];
    expect(copiedCard.id).toBeDefined();
    expect(copiedCard.title).toBe(card.title);
    expect(copiedCard.description).toBe(card.description);
    expect(copiedCard.listId).toBe("2");
    expect(copiedCard.tags).toStrictEqual(card.tags);
    expect(copiedCard.attachments).toStrictEqual(card.attachments);
    expect(stateAfterCopy.modalActionMenusVisibility.value.copyMenuOpen).toBeFalsy();
  });
});
