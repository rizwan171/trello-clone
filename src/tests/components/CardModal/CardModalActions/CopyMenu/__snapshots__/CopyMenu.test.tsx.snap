// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyMenu should render successfully 1`] = `
<DocumentFragment>
  <div
    class="fixed mt-10 w-72 text-gray-700 bg-white rounded-ibsm shadow-2xl p-4"
  >
    <div
      class="relative text-center mb-2"
    >
      <span
        class="text-sm block relative z-10"
      >
        Copy Card
      </span>
      <svg
        class="absolute right-0 top-0 z-20 cursor-pointer"
        data-testid="close-icon"
        fill="currentColor"
        height="20"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
    <hr />
    <div
      class="flex flex-col mt-2 text-sm"
    >
      <span
        class="text-gray-600 my-2 font-semibold"
      >
        Copy To...
      </span>
      <select
        class="rounded-sm p-2 w-full max-w-full overflow-ellipsis"
        name="lists"
      >
        <option
          disabled=""
          hidden=""
          value=""
        >
          Select list
        </option>
        <option
          value="1"
        >
          List 1
        </option>
        <option
          value="2"
        >
          List 2
        </option>
      </select>
      <button
        class="p-2 mt-3 bg-trello-blue-100 hover:bg-trello-blue-200 text-white font-semibold items-center text-sm shadow-sm rounded-sm"
      >
        Copy Card
      </button>
    </div>
  </div>
</DocumentFragment>
`;
